### PR TITLE
Support for parsing multiple internal assets

### PR DESCRIPTION
**Pull request by jaremieromer**
### Goal

Extend NcTools to allow for parsing all meshes from an input file.
If `nc-convert` is called in single-target mode, only the first mesh from an input file will be processed.
If `nc-convert` is called in manifest mode, and legacy syntax is specified for a mesh in the manifest (specifying one asset name), then only the first mesh from an input file will be processed.
If `nc-convert` is called in manifest mode, and multiple asset names are specified for a mesh, then all internal meshes will be processed. Internal meshes that don't exist as specified in the manifest will be skipped to allow for the remaining manifest items to be processed.

### Changes

-  Added flow to allow for parsing of internal meshes (most of this work is done in `Manifest.cpp`)
- Renamed `EnumConversion.h/.cpp` to `EnumExtensions.h/.cpp` and added a method that returns whether the type specified in the enum can output many assets.
- Added test in `NcConvert_integration_tests.cpp` to ensure internal mesh names specified in the manifest will not bring down the whole conversion process if missing. (Will log a warning and skip, instead).
- Modified the docs and the test manifests to allow for the new specification.

### CODE REVIEW

File: source/ncconvert/converters/GeometryConverter.cpp [hunk 1]
```diff
@@ -329,9 +332,34 @@ class GeometryConverter::impl
...
+            {
+                mesh = scene->mMeshes[0];
+            }
+
+            if (mesh->mNumVertices == 0)
+            {
+                throw nc::NcError("No vertices found for: ", path.string());
+            }
+
             auto convertedVertices = ::ConvertToMeshVertices(scene->mMeshes[0]);
```
Comment (McCallisterRomer) [inline source/ncconvert/converters/GeometryConverter.cpp]: Here and in the return statement we're still using `scene->mMeshes[0]` unconditionally, never `mesh`.

File: source/ncconvert/converters/GeometryConverter.cpp [hunk 2]
```diff
@@ -329,9 +332,34 @@ class GeometryConverter::impl
...
         }
 
-        auto ImportMesh(const std::filesystem::path& path) -> asset::Mesh
+        auto ImportMesh(const std::filesystem::path& path, const std::optional<std::string>& internalName) -> asset::Mesh
         {
             const auto scene = ::ReadFbx(path, &m_importer, meshFlags);
+
+            aiMesh* mesh = nullptr;
+
+            if (internalName.has_value())
```
Comment (McCallisterRomer) [inline source/ncconvert/converters/GeometryConverter.cpp]: Consider extracting the mesh selection code into a helper function.

Comment (McCallisterRomer) [reply source/ncconvert/converters/GeometryConverter.cpp]: Not for now, just thinking about the future - I believe Assimp only stores that most recent scene, but I think we could extract and cache them in the converter. Then whenever we go to load a mesh, we first check against the cache to see if we've already parsed that file.

File: source/ncconvert/converters/GeometryConverter.cpp (hunk 1 above)
Comment (jaremieromer) [reply source/ncconvert/converters/GeometryConverter.cpp]: Great catch. Adding a unit test to address this

**Comment by jaremieromer** [comment]
I've also changed "internalName" to "subResourceName", and removed the exception checking swallow in Builder.cpp

File: source/ncconvert/converters/GeometryConverter.cpp [hunk 3]
```diff
@@ -48,17 +48,37 @@ auto ReadFbx(const std::filesystem::path& path, Assimp::Importer* importer, unsi
...
         throw nc::NcError("Fbx contains no mesh data\n    file: ", path.string());
     }
-    else if (scene->mNumMeshes > 1)
+
+    return scene;
+}
+
+auto GetMeshFromScene(const aiScene* scene, const std::optional<std::string>& subResourceName = std::nullopt) -> aiMesh*
+{
+    aiMesh* mesh = nullptr;
```
Comment (McCallisterRomer) [inline source/ncconvert/converters/GeometryConverter.cpp:57]: IMO this function would read a little easier if you swapped the cases - first check if no subresource -> return first mesh. 

```cpp
NC_ASSERT(scene->mNumMeshes, "...");
if (!subResourcename.has_value())
{
   return scene->mMeshes[0];
}

// Other code, but slightly less indented
```

Comment (McCallisterRomer) [reply source/ncconvert/converters/GeometryConverter.cpp:57]: Dealers choice between assert and the if/throw you have. NC_ASSERT is on (at least here) even in release builds.

File: source/ncconvert/utility/EnumExtensions.cpp [hunk 4]
```diff
@@ -7,6 +7,31 @@
...
+            return false;
+        case asset::AssetType::ConcaveCollider:
+            return false;
+        case asset::AssetType::HullCollider:
+            return false;
+        case asset::AssetType::Mesh:
+            return true;
+        case asset::AssetType::Texture:
+            return false;
+        default:
```
Comment (McCallisterRomer) [inline source/ncconvert/utility/EnumExtensions.cpp:26]: Was the compiler complaining about no default? IIRC, if you leave the default out, you can get a warning if you add a new enum value and forget to add it to the switch (from gcc maybe?).

Comment (McCallisterRomer) [reply source/ncconvert/utility/EnumExtensions.cpp:26]: Ahh `ToString` below has default too.

**Review by McCallisterRomer — APPROVED** [review]

**Pull request merged**